File: Game-1-modular/core/effect_context.py
"""
Effect Context and Configuration
Defines the data structures for effect execution
"""

from dataclasses import dataclass, field
from typing import List, Dict, Any, Optional


@dataclass
class EffectConfig:
    """
    Configuration for an effect parsed from tags
    This is the unified structure all effects use
    """
    # Source tags
    raw_tags: List[str] = field(default_factory=list)

    # Categorized tags
    geometry_tag: Optional[str] = None
    damage_tags: List[str] = field(default_factory=list)
    status_tags: List[str] = field(default_factory=list)
    context_tags: List[str] = field(default_factory=list)
    special_tags: List[str] = field(default_factory=list)
    trigger_tags: List[str] = field(default_factory=list)

    # Resolved context
    context: str = "enemy"  # "ally", "enemy", "self", "all", etc.

    # Base parameters
    base_damage: float = 0.0
    base_healing: float = 0.0

    # All effect parameters (merged defaults + user params)
    params: Dict[str, Any] = field(default_factory=dict)

    # Warnings/conflicts
    warnings: List[str] = field(default_factory=list)
    conflicts_resolved: List[str] = field(default_factory=list)


@dataclass
class EffectContext:
    """
    Runtime context for effect execution
    Contains source, target(s), and config
    """
    source: Any  # Entity that created the effect
    primary_target: Any  # Primary target entity
    config: EffectConfig  # Effect configuration
    timestamp: float = 0.0  # When effect was created
    targets: List[Any] = field(default_factory=list)  # All targets (after geometry)

    def __post_init__(self):
        if not self.targets and self.primary_target:
            self.targets = [self.primary_target]

File: Unity/Assets/Scripts/Game1.Systems/Tags/EffectContext.cs
// Game1.Systems.Tags.EffectContext
// Migrated from: core/effect_context.py (EffectContext dataclass)
// Migration phase: 4
//
// Runtime context for effect execution.
// Contains source, target(s), and config.

using System.Collections.Generic;

namespace Game1.Systems.Tags
{
    /// <summary>
    /// Runtime context for effect execution.
    /// Created by EffectExecutor, populated during execution with
    /// resolved targets and timing information.
    /// </summary>
    public class EffectContext
    {
        /// <summary>Entity that created the effect (caster, turret, etc.).</summary>
        public object Source { get; set; }

        /// <summary>Primary target entity before geometry resolution.</summary>
        public object PrimaryTarget { get; set; }

        /// <summary>Parsed effect configuration (tags, params, context).</summary>
        public EffectConfig Config { get; set; }

        /// <summary>Timestamp when the effect was created (game time).</summary>
        public float Timestamp { get; set; } = 0f;

        /// <summary>All resolved targets after geometry selection.</summary>
        public List<object> Targets { get; set; } = new();

        /// <summary>
        /// Creates a new EffectContext. Mirrors Python __post_init__:
        /// if Targets is empty and PrimaryTarget is not null,
        /// Targets is initialized to [PrimaryTarget].
        /// </summary>
        public EffectContext(object source, object primaryTarget, EffectConfig config,
                             float timestamp = 0f, List<object> targets = null)
        {
            Source = source;
            PrimaryTarget = primaryTarget;
            Config = config;
            Timestamp = timestamp;
            Targets = targets ?? new List<object>();

            // Python __post_init__ equivalent
            if (Targets.Count == 0 && PrimaryTarget != null)
            {
                Targets.Add(PrimaryTarget);
            }
        }
    }
}
